Ability to update core URL without mf url change (#1111)
SAP/luigi: docs/tutorials.md

<!-- meta
{
  "node": {
    "label": "Tutorials",
    "category": {
      "label": "Basics"
    },
    "metaData": {
      "categoryPosition": 1,
      "position": 3
    }
  }
}
meta -->

# Tutorials

> A collection of advanced usecases and example implementations. If you are new to Luigi, you might want to have a look at our [**Getting Started**](getting-started.md) section first or one of our SPA implementations in the **Examples** section on the left.

Reviewer: ```suggestion
This is a collection of advanced use cases and example implementations. If you are new to Luigi, you might want to have a look at our [Getting Started](getting-started.md) section first or one of our "Hello World" SPA implementations in the **Examples** section on the left.
```

<!-- accordion:start -->

### Use a SPA router for navigation and keep Luigi Core URL in sync

> It can be convenient to keep an existing routing strategy, and be able to use an existing microfrontend as drop-in without the need to refactor everyting to `LuigiClient.linkManager()`. 

Reviewer: ```suggestion
It can be convenient to keep an existing routing strategy, and be able to use an existing micro frontend as drop-in without the need to refactor everything to `LuigiClient.linkManager()`. 
```

> Concept wise it is neccessary to update the Luigi Core URL when routing internally with the microfrontend router, but without an update of the URL on Luigi Client side. This can be achieved by using the `linkManager()` [**withoutSync**](luigi-client-api.md#withoutsync) method.

Reviewer: ```suggestion
It is necessary to update the Luigi Core URL when routing internally with the micro frontend router, but without an update of the URL on Luigi Client side. This can be achieved by using the `linkManager()` [withoutSync](luigi-client-api.md#withoutsync) method.
```

> The following example shows how to configure the navigation node and use router events for automatic update of the url when using the microfrontends built in router.

Reviewer: ```suggestion
The following example shows how to configure the navigation node and use router events for automatic update of the URL when using the micro frontend's built in router.
```

> Be aware that this is a very simple example for navigation and special cases like modals, split views and other functionalities still require to use `LuigiClient`.

Reviewer: ```suggestion
Be aware that this is a very simple example for navigation. Special cases like modals, split views and other functionalities still require the use of the Luigi Client.
```

> For keeping the example simple, we use [**virtualTree**](navigation-parameters-reference.md#virtualtree) to allow any nested navigation. This is not mandatory, you can always specify the node tree by yourself and still use automatic navigation with router events.

Reviewer: ```suggestion
To keep the example simple, we use [virtualTree](navigation-parameters-reference.md#virtualtree) to allow any nested navigation. This is not mandatory; you can always specify the node tree by yourself and still use automatic navigation with router events.
```

#### Luigi node configuration

```javascript
    {
      pathSegment: 'Orders',
      label: 'orders',
      viewUrl: 'https://orders.microfrontend.tld/',
      navigationContext: 'orders',
      virtualTree: true
    }
```

#### Use Angular Router for navigation

> Angular provides Router events. We are reacting on NavigationEnd to update the URL after a successful route change.

Reviewer: ```suggestion
Angular provides [Router events](https://angular.io/guide/router#router-events). We are reacting on `NavigationEnd` to update the URL after a successful route change.
```

Reviewer: If you think the link is unnecessary here we can remove it. This is just how it's usually done in Kyma documentation 😅

> We assume that the whole Angular app is one microfrontend and has its routes declared on the root level:

Reviewer: ```suggestion
We assume that the whole Angular app is one micro frontend and has its routes declared on the root level:
```

  { path: ':id/details', component: OrderDetailsComponent },
```

Implement `luigi-auto-navigation.service.ts`, which is globally imported in our `app.module.ts`:
```javascript
    import { Router, NavigationEnd } from '@angular/router';
    import { Injectable, OnDestroy } from '@angular/core';
    import { Subscription } from 'rxjs';
    import { filter } from 'rxjs/operators';
    import { linkManager } from '@kyma-project/luigi-client';

    @Injectable()
    export class LuigiAutoNavigationService implements OnDestroy {
      private subscriptions: Subscription = new Subscription();
      constructor(private router: Router) {}
      public init(): void {
        this.subscriptions.add(
          this.router.events
            .pipe(filter(ev => ev instanceof NavigationEnd))
            .subscribe((ev: NavigationEnd) => {
              if (ev instanceof NavigationEnd) {
                linkManager()
                  .fromClosestContext()
                  .withoutSync()
                  .navigate(ev.url);
              }
            })
        );
      }

      ngOnDestroy(): void {
        this.subscriptions.unsubscribe();
      }
    }
```

> Other than adding the service, which also could be implemented as a RouteGuard or similar, the microfrontend is unchanged and uses `[routerLink='']` or other functionalities to navigate.

Reviewer: ```suggestion
Other than adding the service, which also could be implemented as a `RouteGuard` or similar, the micro frontend is unchanged and uses `[routerLink='']` or other functionalities to navigate.
```
